### ISSUE TITLE
tensorflow_ranking.keras.losses.ListMLELoss() gives -inf 

### ISSUE THREAD

**Issue by mrhhyu** [tensorflow/tensorflow#56749]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Fedpea 36

### Current Behaviour?

```shell
In 'tensorflow_ranking' (i.e., in short tfr) package, 
tfr.keras.losses.ListMLELoss() produces 'inf' result for float numbers with large decimals since TF converts them to 0. 
I spent 2 days to find the reason; in the ListMLELoss() source code, line 1304 
{sums = tf.math.log(sums) - sorted_logits}
when sums is a floating number with large decimals, it is counted as zero and the result would be inf. 
We can change the aforementioned line as follows to have numerical stability:
{sums = tf.math.log(sums + tf.keras.backend.epsilon()) - sorted_logits}
Also, tfr.keras.metrics.NDCGMetric() suffers from similar problem as well but for large numbers (implementation of Normalized Discounted Cumulative Gain is so complicated in TF.tfr); sklearn gives the correct answer and has clear implementation, while TF gives Nan.!!
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf 
import tensorflow_ranking as tfr
from sklearn.metrics import ndcg_score

y_true = [np.arange(0,1000, dtype='float32')]
y_pred = [np.arange(0,1000, dtype='float32')]

loss = tfr.keras.losses.ListMLELoss()
print('loss: ', loss(y_true, y_pred).numpy() )

ndcg = tfr.keras.metrics.NDCGMetric()
print('ndcg by tensorflow: ',ndcg(y_true, y_pred))

print('ndcg by skilear:', ndcg_score(y_true, y_pred))
```

### Relevant log output

```shell
2022-07-13 16:00:41.562605: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2022-07-13 16:00:41.562634: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: XX
2022-07-13 16:00:41.562640: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: XX
2022-07-13 16:00:41.562688: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 510.68.2
2022-07-13 16:00:41.562705: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 510.68.2
2022-07-13 16:00:41.562710: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:310] kernel version seems to match DSO: 510.68.2
2022-07-13 16:00:41.562899: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
loss:  -inf
ndcg by tensorflow:  tf.Tensor(nan, shape=(), dtype=float32)
ndcg by skilear: 0.9999999999999998
```
</details>

**Comment by sushreebarsa** [comment]
@gadagashwini I was able to replicate the issue on colab, please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/46b093f17db92755410a284b1d7ad8e5/56749.ipynb#scrollTo=1Bh6MTyrlH9i). Thank you!

**Comment by mrhhyu** [comment]
> @gadagashwini I was able to replicate the issue on colab, please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/46b093f17db92755410a284b1d7ad8e5/56749.ipynb#scrollTo=1Bh6MTyrlH9i). Thank you!

Thank you, the problems are raised due to the wrong implementations of ListMLE and DCGMetric in tensorflow_ranking.

**Comment by sachinprasadhs** [comment]
@ramakumar1729 , Could you please look into the issue. Thanks!

**Comment by mrhhyu** [comment]
Unfortunately, most of the loss functions (e.g., ListMLELoss, UniqueSoftmaxLoss, GumbelApproxNDCGLoss) and metrics in tensorflow_ranking package have the same problem when log(0) for 2^n (for large n).

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
